packages/common/src/tools/queue/TaskQueue.ts

```diff
@@ -0,0 +1,113 @@
+import assert from 'assert'
+
+import { Logger } from '../Logger'
+import { Retries } from './Retries'
+import { Job, ShouldRetry } from './types'
+
+interface TaskQueueOpts<T> {
+  workers?: number
+  shouldRetry?: ShouldRetry<T>
+}
+
+export class TaskQueue<T> {
+  private queue: Job<T>[] = []
+  private busyWorkers = 0
+  private workers: number
+  private shouldRetry: ShouldRetry<T>
+
+  constructor(
+    private executeTask: (task: T) => Promise<void>,
+    private logger: Logger,
+    opts?: TaskQueueOpts<T>,
+  ) {
+    this.workers = opts?.workers ?? 1
+    assert(
+      this.workers > 0 && Number.isInteger(this.workers),
+      'workers needs to be a positive integer',
+    )
+    this.shouldRetry =
+      opts?.shouldRetry ??
+      Retries.exponentialBackOff(100, { maxAttempts: 10, maxDistance: 10_000 })
```

Comment:
If we include maxAttempts as a default the current system would break because there is an assumption that tasks will be finished eventually. Also 10s is quite a short time, maybe we could have 10 minutes instead?

Reply:
That's why I set `always` as default in the first place. I think for the purpose of delivery I'd keep `always` as default and have it thought through as a separate PR. Wdyt?

Reply:
I agree. Let's have always for now